### ISSUE TITLE
Coord-check for conv1d

### ISSUE THREAD

**Issue by bob80333** [microsoft/mup#14]
I modified the muconv2d branch to get a conv1d variant of the output layer for mup, and I applied it to a shallow variant of a unet model I've been testing.

repo for model: https://github.com/bob80333/audio_bandwidth_extension
fork of mup with conv1d: https://github.com/bob80333/mup/tree/muconv2d

Here's the coord-check results, they don't look quite a smooth as the paper but there's definitely a big difference between mup and no mup.

mup:
![plot_mup](https://user-images.githubusercontent.com/6364377/168668767-5fb11f2e-085d-442b-bb54-4a6ba706c9be.png)

no mup:
![plot_nomup](https://user-images.githubusercontent.com/6364377/168668789-a7de33b7-f646-4f22-bd47-fe9162aa4292.png)

Does this look about right for the coordinate check?  The figures I saw in the example looked much smoother than this.

**Comment by edwardjhu** [comment]
Hi Bob,

Thanks for sharing these! The mup plots look off. You can see the blowup in width in the last layer. I looked through your code and couldn't spot anything obvious. Can you try kernel_size=1, which should reduce it to just mup.Linear? We can see if it behaves the same as mup.Linear and start from there.

**Comment by bob80333** [comment]
Just to clarify, only the output (MuReadout style) conv1d layer should have kernel_size=1?

**Comment by edwardjhu** [comment]
Yup, just the output layer.

**Comment by bob80333** [comment]
kernel size 1:

mup:
![plot_mup_outkernel1](https://user-images.githubusercontent.com/6364377/168704313-00fbe582-ea27-4f11-a405-9fd50a253336.png)

no mup:
![plot_nomup_outkernel1](https://user-images.githubusercontent.com/6364377/168704325-95929f89-14da-4669-8371-86950da4a646.png)

**Comment by edwardjhu** [comment]
Can you also try mup.Linear? It should give you the same result if mup.Conv1D is implemented correctly. If the curves still look off, then either there's a bug in mup.Linear or there's a problem in how mup is used, barring some other factors in your model.

**Comment by bob80333** [comment]
I tried a few things.

There is a skip connection in this model between the output and the input, so I tried replacing that Conv1d with mup Conv1d.

Result:
mup:
![plot_mup_outkernel1_lastskip](https://user-images.githubusercontent.com/6364377/168884170-024c1d04-5654-47a3-9cb5-266d78cdc681.png)
no mup:
![plot_nomup_outkernel1_lastskip](https://user-images.githubusercontent.com/6364377/168884202-38821205-1ddf-4375-9269-8f596b6ee25f.png)

Replacing the out_conv with a linear layer, leaving the skip connection as it was originally:
mup:
![plot_mup_outlinear](https://user-images.githubusercontent.com/6364377/168884336-c6dfe31c-b69c-4b4e-8ae2-4f8d80389e4a.png)
no mup:
![plot_nomup_outlinear](https://user-images.githubusercontent.com/6364377/168884360-213c33b1-52be-4e95-a5fd-8d1707634a26.png)

Replacing the skip connection and out_conv with linear layers (MupReadout)
mup:
![plot_mup_outlinear_lastskip](https://user-images.githubusercontent.com/6364377/168884472-ae3a112d-4b94-417d-bd69-969f1b9a7349.png)
no mup:
![plot_nomup_outlinear_lastskip](https://user-images.githubusercontent.com/6364377/168884489-613c2a34-1550-43f2-b7c6-cb7cc8153a08.png)

**Comment by bob80333** [comment]
Removing weight norm only from output layer (mup Linear readout layer)

with mup:
![plot_mup_outlinear_removewnout](https://user-images.githubusercontent.com/6364377/169423631-77ccf863-313e-4aca-aae7-04e720f9f83e.png)
without mup:
![plot_nomup_outlinear_removewnout](https://user-images.githubusercontent.com/6364377/169423642-57223874-36fc-4d16-8148-53c1725afed1.png)

Removing weight norm from all layers (mup Linear readout layer)

with mup:
![plot_mup_outlinear_removewnall](https://user-images.githubusercontent.com/6364377/169423674-f15b0ae3-bfb7-4c17-b0f1-cf1086e78af8.png)
without mup:
![plot_nomup_outlinear_removewnall](https://user-images.githubusercontent.com/6364377/169423685-cace171b-62ec-4f93-b9c6-faf823af49d2.png)

**Comment by thegregyang** [comment]
@bob80333 there seems to be some slight blow up issue at initialization already actually. To speed this up, perhaps you can recreate this in a colab notebook so I can debug it myself and get back to you?

**Comment by bob80333** [comment]
@thegregyang sorry for the delay, here's a colab notebook that should work:

https://colab.research.google.com/drive/1OvQ_My8KOOUowtrY56oQm7PFqNOn2hbS?usp=sharing

**Comment by bob80333** [comment]
@thegregyang Did the colab notebook work for you?

**Comment by thegregyang** [comment]
@bob80333 I have found the issue, which has to do with the modification you made to your fork of `mup`. I'll reply today with a working notebook.

**Comment by thegregyang** [comment]
Here's [a working colab](https://colab.research.google.com/drive/1iLAnPM33Ie5A8i2QYWqgXeFr0Ml-6lSZ?usp=sharing). Coord check seems to work for both the shallow net and your original deep net.

Shallow:
![image](https://user-images.githubusercontent.com/53244851/172074249-2c7961d9-1f8d-459a-a093-65b11f2d85a8.png)

Deep:
![image](https://user-images.githubusercontent.com/53244851/172074245-4a0a3fd2-dcde-4b01-8cd3-918eb4716d66.png)

At least one of the issues was [your modification here](https://github.com/bob80333/mup/blob/muconv2d/mup/shape.py#L194) which did not rescale Conv's initialization correctly. After that I switched back to our original branch and the coord check essentially passed without problems after adding in the Conv1d layer.

**Comment by bob80333** [comment]
Awesome! Thank you!

I was playing around with trying to get weight norm working, and I think I had an idea to make it work.

The pytorch documentation about weight norm from [here](https://pytorch.org/docs/stable/generated/torch.nn.utils.weight_norm.html) says:

`
This replaces the parameter specified by name (e.g. 'weight') with two parameters: one specifying the magnitude (e.g. 'weight_g') and one specifying the direction (e.g. 'weight_v').
`

Given that we are rescaling the parameters, I think we could modify this line:

https://github.com/microsoft/mup/blob/16ef49056825daaa7bc7c691024bd66586a9f6a6/mup/layer.py#L123

and add a check if 'weight_g' (The direction of the weight for weight norm) exists, and use that instead of weight.

If I remember correctly, this was actually the reason behind my modification ('weight' didn't exist since I used weight norm, so I changed it get the infshape from the bias (which was incorrect)).  

Since 'weight_g' is the direction of the weight vector, the shape is the same as the original 'weight' parameter (I checked), so I assume the infshape should be the same as well.

**Comment by bob80333** [comment]
@thegregyang I tried out a modification to get weight norm working, not sure if it's correct.

WN + MuP, full model depth:

![download (1)](https://user-images.githubusercontent.com/6364377/172458325-a0298522-00e3-4fe6-a201-ede205829a5a.png)

Here's a colab with the code:

https://colab.research.google.com/drive/1osnDz7dRu8Y86V6klX02EA60PLULRGLr?usp=sharing

**Comment by thegregyang** [comment]
> @thegregyang I tried out a modification to get weight norm working, not sure if it's correct.
> 
> WN + MuP, full model depth:
> 
> ![download (1)](https://user-images.githubusercontent.com/6364377/172458325-a0298522-00e3-4fe6-a201-ede205829a5a.png)
> 
> Here's a colab with the code:
> 
> https://colab.research.google.com/drive/1osnDz7dRu8Y86V6klX02EA60PLULRGLr?usp=sharing

The coord check doesn't look good.

The problem with weight-norm is not the code but the mathematics. The way weight-norm normalizes hidden weight matrix is not natural: one should normalize it to Frobenius norm $\Theta(\sqrt{width})$ instead of $\Theta(1)$. A lot of things need to be fixed to get weight-norm to scale correctly.

Would you be OK with not using weight-norm? You can just replace it with layernorm or batchnorm.

**Comment by bob80333** [comment]
Thanks for the explanation.  I tried multiple different norm layers, and I found that for my model the only norm that didn't hurt performance was weightnorm, so I removed all normalization.
